### PR TITLE
LPS Network Endpoint Enhancements and Signaling

### PR DESCRIPTION
**Pull request by milan-zededa**
# Description

Implements the EVE side of the API additions introduced in
[lf-edge/eve-api#144](https://github.com/lf-edge/eve-api/pull/144).

Three API enhancements plus two ancillary improvements:

### 1. Signal endpoint — low-latency LPS config notifications (`GET /api/v1/signal`)

EVE now maintains a persistent long-lived HTTP GET connection to the optional
`/api/v1/signal` endpoint on the Local Profile Server.  LPS can push a
`Signal` proto message (NDJSON-framed) listing which endpoints have a pending
configuration change.  On receipt, EVE immediately polls the listed endpoints
instead of waiting for the next scheduled tick.

Implementation details:
- New goroutine in `pkg/pillar/localcommand/signal.go` handles the stream
  lifecycle independently of the watchdog.
- TCP keepalive at 60 s is the sole liveness mechanism (no app-layer heartbeat),
  as required by the spec.
- Exponential backoff on failure (1 s initial, 2× doubling, 30 s cap).
- If LPS returns 404 (endpoint not implemented), reconnect attempts are throttled
  to one per hour so a non-signaling LPS does not generate excess traffic.
- Rate limiter (one signal per 3 s, burst 3) guards against a buggy or malicious
  LPS overwhelming EVE with spurious triggers.  Periodic polling is the
  correctness guarantee; dropped signals are safe.
- The stream is restarted whenever the LPS address changes (`UpdateLpsConfig`).
- Unknown `ConfigEndpoint` enum values are silently ignored for forward
  compatibility.

### 2. Per-port runtime status (`NetworkInfo.port_status`)

EVE now populates the new `NetworkPortStatus` repeated field in every
`POST /api/v1/network` request.  For each network port it reports the
kernel-observed runtime state: link up/down, MAC address, assigned IP
addresses (CIDR), active default gateways, DNS servers, DNS search domain,
NTP servers, and MTU.

### 3. `local_modifications_allowed` per port in `NetworkInfo.latest_config`

EVE sets `NetworkPortConfig.local_modifications_allowed` when reporting
`latest_config` in `NetworkInfo`, mirroring the controller-provisioned
`SystemAdapter.allow_local_modifications` flag.  LPS can use this to know
upfront which ports accept locally submitted configuration, rather than
discovering it via a trial-and-error error message.

### 4. Reactive LPS network POST on config/status change

EVE now triggers an immediate `POST /api/v1/network` when either:
- `DevicePortConfigList` changes (new network configuration applied), or
- `DeviceNetworkStatus` changes with a meaningful state update (link state,
  IP assignment, etc.).

Previously the network endpoint was only driven by the periodic ticker.

### 5. Configurable LPS polling intervals

All six LPS polling intervals are now tunable via controller config
properties (previously they were compile-time constants):

| Property                         | Default | Min | Max  |
|----------------------------------|---------|-----|------|
| `timer.lps.profile.interval`     | 60 s    | 3 s | 1 h  |
| `timer.lps.radio.interval`       | 5 s     | 3 s | 1 h  |
| `timer.lps.appinfo.interval`     | 60 s    | 3 s | 1 h  |
| `timer.lps.devinfo.interval`     | 60 s    | 3 s | 1 h  |
| `timer.lps.network.interval`     | 60 s    | 3 s | 1 h  |
| `timer.lps.appbootinfo.interval` | 60 s    | 3 s | 1 h  |

Defaults match the previous hard-coded values, so behaviour is unchanged
without explicit configuration.

### Prerequisites

You need a running EVE device with a deployed Local Profile Server application.
Consider using [my LPS implementation](https://github.com/milan-zededa/eve/tree/evetest/evetest/testapps/lps).

### 1. Validate `local_modifications_allowed` in NetworkInfo

**LPS side:** In the handler for `POST /api/v1/network`, inspect each entry in
`network_info.latest_config.ports`.  Ports provisioned by the controller with
`allow_local_modifications = true` must arrive with
`local_modifications_allowed = true`; those provisioned without it must arrive
with `false`.

**Expected behaviour:** LPS can display a read-only indicator next to ports
that it is not permitted to reconfigure, and can pre-validate a
`LocalNetworkConfig` submission before sending it.

### 2. Validate per-port runtime status (`port_status`)

**LPS side:** In the same `POST /api/v1/network` handler, inspect
`network_info.port_status`.  For each network port you should see:
- `logical_label` — matches the port label from `latest_config`
- `interface_name` — kernel interface name (e.g., `eth0`)
- `link_up` — `true` when the port has carrier
- `mac_address` — colon-separated hex (e.g., `aa:bb:cc:dd:ee:ff`)
- `ip_addresses` — CIDR notation (e.g., `192.168.1.10/24`, `fd00::1/64`)
- `gateways` — active default routers
- `dns_servers` / `dns_domain` — effective resolver configuration
- `ntp_servers` — NTP peers in use
- `mtu` — effective MTU in bytes

**To verify reactivity (enhancement 4):** Change a network configuration on
the device (e.g., add a static route, trigger a DHCP renewal).  LPS should
receive an updated `POST /api/v1/network` within a few seconds, not after the
next 60-second tick.

### 3. Validate the Signal endpoint — basic trigger

**LPS side:** Implement `GET /api/v1/signal` as a streaming endpoint that:
1. Keeps the HTTP connection open indefinitely (use chunked transfer encoding).
2. When a user submits a configuration change (e.g., via a UI that calls
   `PUT /api/v1/local_profile`), immediately writes a single NDJSON line:
   ```
   {"pendingChanges":["CONFIG_ENDPOINT_LOCAL_PROFILE"]}\n
   ```
   and flushes the response buffer.
3. Does NOT write anything until there is actually a pending change.
4. Does NOT write heartbeats or empty lines.

**EVE side:** After receiving the signal, EVE must immediately fetch
`GET /api/v1/local_profile` and apply the new profile — without waiting for
the next 60-second tick.

**Verification:** Measure the latency between submitting the config change on
LPS and EVE applying it.  With signaling it should be under 2 seconds; without
signaling it would be up to 60 seconds.

### 4. Validate the Signal endpoint — multi-endpoint coalescing

Send a signal listing multiple endpoints simultaneously:
```
{"pendingChanges":["CONFIG_ENDPOINT_NETWORK","CONFIG_ENDPOINT_APP_INFO"]}\n
```
EVE must trigger both the network POST and the app info POST immediately.

### 5. Validate configurable polling intervals

Using the controller, set:
```
timer.lps.profile.interval = 10
```
Observe in EVE logs that the local profile is now fetched every 10 seconds
instead of every 60 seconds.  Restore the default when done.

## Changelog notes

- **Low-latency LPS configuration updates.** EVE can now receive near-instant
  notifications from the Local Profile Server when new configuration is ready,
  reducing the delay from up to 60 seconds to under 2 seconds.  LPS
  applications must implement the optional `GET /api/v1/signal` streaming
  endpoint to take advantage of this; existing LPS deployments without the
  endpoint continue to work exactly as before.

- **Richer network status in LPS.** EVE now reports the live kernel-observed
  state of every network port to LPS on each network POST: link up/down,
  assigned IP addresses (with subnet mask), default gateways, DNS and NTP
  servers, and MTU.  LPS applications can display this information to operators
  without needing a separate management channel.

- **Per-port modification permission flag.** LPS now knows which network ports
  it is allowed to reconfigure (as provisioned by the controller), so it can
  give operators clear feedback instead of a generic error when they attempt to
  modify a controller-managed port.

- **Reactive LPS network updates.** EVE now pushes an updated network POST to
  LPS immediately when the device's network configuration or link state changes,
  rather than waiting for the next periodic tick.

- **Tunable LPS polling intervals.** All six LPS polling intervals (local
  profile, radio, app info, device info, network, app boot info) can now be
  adjusted via controller config properties.  The defaults are unchanged.

## PR Backports

```text
- 16.0-stable: Yes. Even though this is strictly speaking a new feature, we must make an exception here...
- 14.5-stable: No 
- 13.4-stable: No
```
## Checklist

- [x] I've provided a proper description
- [x] I've added the proper documentation
- [x] I've tested my PR on amd64 device
- [ ] I've tested my PR on arm64 device
- [x] I've written the test verification instructions
- [x] I've set the proper labels to this PR
- [x] I've checked the boxes above, or I've provided a good reason why I didn't check them.

### CODE REVIEW

**Review by eriknordmark — APPROVED** [review]
Run tests

But some yetus annotations to fix (however, yetus shows as passing which is new to me!)

**Comment by rene** [comment]
> Run tests
> 
> But some yetus annotations to fix (however, yetus shows as passing which is new to me!)

If you look to the artifacts, Yetus is 100% passing because it checks only the patch (what it was changed), so if you go to the final result files, they are all blank (no errors).

However, the annotation step runs over the `patch-*-result` files, which shows all errors for the file and not only what was changed.... the annotation step will not fail (I did it on purpose). I think we can keep this approach while we solve the issues as they show up on files we are changing...

But it's important to point out that any error introduced by the PR will make Yetus to fail.... so the check is working.....

**Comment by milan-zededa** [comment]
> > Run tests
> > But some yetus annotations to fix (however, yetus shows as passing which is new to me!)
> 
> If you look to the artifacts, Yetus is 100% passing because it checks only the patch (what it was changed), so if you go to the final result files, they are all blank (no errors).
> 
> However, the annotation step runs over the `patch-*-result` files, which shows all errors for the file and not only what was changed.... the annotation step will not fail (I did it on purpose). I think we can keep this approach while we solve the issues as they show up on files we are changing...
> 
> But it's important to point out that any error introduced by the PR will make Yetus to fail.... so the check is working.....

I think this approach is quite annoying for the reviewer. Furthermore, these are not "issues", but suggestions. We may not agree with everything pointed out by Yetus, so we will intentionally ignore some, but they will keep popping up...

**Comment by eriknordmark** [comment]
> However, the annotation step runs over the `patch-*-result` files, which shows all errors for the file and not only what was changed.... the annotation step will not fail (I did it on purpose). I think we can keep this approach while we solve the issues as they show up on files we are changing...

This is at best confusing (because those annotations show up in the diffs the same way as when they cause failures) and a bit annoying. Can we have the annotations be limited to the changed lines as before.

**Review by uncleDecart — APPROVED** [review]

**Comment by milan-zededa** [comment]
Rebased on the top of the latest master.

File: docs/CONFIG-PROPERTIES.md [hunk 1]
```diff
@@ -96,6 +96,12 @@
...
+| timer.lps.profile.interval | integer in seconds | 60 (1 minute) | 3 | 3600 (1 hour) | how frequently EVE fetches the local profile from the Local Profile Server (LPS) |
+| timer.lps.radio.interval | integer in seconds | 5 | 3 | 3600 (1 hour) | how frequently EVE POSTs radio status to LPS and fetches radio silence configuration |
+| timer.lps.appinfo.interval | integer in seconds | 60 (1 minute) | 3 | 3600 (1 hour) | how frequently EVE POSTs application info to LPS and fetches application commands |
+| timer.lps.devinfo.interval | integer in seconds | 60 (1 minute) | 3 | 3600 (1 hour) | how frequently EVE POSTs device info to LPS and fetches device commands |
+| timer.lps.network.interval | integer in seconds | 60 (1 minute) | 3 | 3600 (1 hour) | how frequently EVE POSTs network configuration to LPS and fetches locally-made network configuration |
+| timer.lps.appbootinfo.interval | integer in seconds | 60 (1 minute) | 3 | 3600 (1 hour) | how frequently EVE POSTs application boot info to LPS and fetches boot configuration |
```
Comment (eriknordmark) [inline docs/CONFIG-PROPERTIES.md:109]: Does it make sense to add a general note about all these below the table that if the device gets a 404(??) it will back off the timer a lot?

**Review by eriknordmark — COMMENTED** [review]
If/when the long-lived connection dies and is re-established is the re-establishment done automatically by EVE?

Who is responsible for ensuring that the LPS can track the current state after the re-establishment?
 Should the LPS notice that there is a new TCP/HTTPS connection and use that to trigger fetching/updating the current state? If so, would be be useful if EVE sent a signal over the long-lived connection immediately after it has been re-established the connection so the LPS doesn't need to detect new connections at the lower layers?

**Comment by milan-zededa** [comment]
> If/when the long-lived connection dies and is re-established is the re-establishment done automatically by EVE?
> 
> Who is responsible for ensuring that the LPS can track the current state after the re-establishment? Should the LPS notice that there is a new TCP/HTTPS connection and use that to trigger fetching/updating the current state? If so, would be be useful if EVE sent a signal over the long-lived connection immediately after it has been re-established the connection so the LPS doesn't need to detect new connections at the lower layers?

This signal endpoint is really just an optimization to minimize latency. If we loose some signal from LPS for whatever reason, the only consequence is just longer latency (like it is today without this enhancement).
So we do not impose that LPS must react to re-connection in any way. A good implementation could track all pending changes and re-emit them when reconnection is detected (a new GET request is received for /signal). But a "dumb" implementation which does not keep track of pending changes and simply sends&forget signal on any user input will functionality-wise work as well (even though user-experience is not the best in these edge-cases).

**Review by eriknordmark — COMMENTED** [review]
A few comments:

  1. No unit tests for signal.go. A new ~300-line state machine with five       
  outcomes, an exported public API (OpenLocalStream), and intricate
  cancel/restart semantics ships without a single test. At minimum, handleSignal
   (rate-limit + dispatch table) and readSigHandlerStream (NDJSON framing +
  malformed-line tolerance) are easy to unit-test with an io.Pipe. Worth adding
  before merge given the long-lived-connection complexity.

  2. sigHandlerLimiter config inconsistency. Burst is 3 over a 3 s refill, so a
  single legitimate signal listing all 6 endpoints will not exceed the limit (it
   dispatches synchronously inside handleSignal and consumes 1 token, not 6 —
  Allow() is called once per signal, not per endpoint). That is fine for normal 
  use, but the comment says "periodic polling is the correctness guarantee;
  dropped signals are safe" while the limiter is actually quite generous.
  Consider documenting that the limiter is per-message, not per-endpoint, so
  future maintainers don't tighten it incorrectly.

  3. TCP keepalive is key for the robustness, but if we fail to set it there isn't even any logs in. Makes sense adding such logs.
  (controllerconn/send.go:111-115):                                             
  if tcpConn, ok := conn.(*net.TCPConn); ok {
      _ = tcpConn.SetKeepAlive(true)                                            
      _ = tcpConn.SetKeepAlivePeriod(keepAlive)                                 
  }

  3. If DialerWithResolverCache ever wraps the connection (e.g., for proxying), 
  keepalive — the only dead-peer detection mechanism per the design — is        
  silently disabled. At least log at Trace level when the assertion fails so
  this doesn't become a silent debugging problem later.                         

  4. Empty mask → no /N in CIDR string. dnsPortsToProto builds &net.IPNet{IP: 
  addr.Addr, Mask: addr.Mask} and calls String(). If addr.Mask is nil/empty     
  (possible on a freshly persisted DPC before updateDNS has run, or if a code
  path forgets to populate it), IPNet.String() returns just the IP — violating  
  the proto comment that promises CIDR format. Either guarantee Mask presence by
   construction, or fall back to /32 or /128 based on len(addr.Addr) so the wire
   format is always CIDR.

  5. Log line could be large (signal.go:864): Warnf(... line=%q) logs up to 64
  KiB of attacker-controlled bytes per malformed line. Truncate to ~256 chars in
   the format string.

  6. Unbounded pendingChanges slice.
  A malicious LPS can send a Signal with pendingChanges containing millions of
  entries (within the 64 KiB line cap, that's roughly 2k–5k enum repetitions).  
  Each is iterated in handleSignal. Iteration is O(n) with cheap per-element
  work (one switch + one tickNow), so a 5k-entry signal completes in            
  microseconds. Not exploitable, but a len(pendingChanges) > 32 early-return
  would cost nothing.

7. add a one-line // see eve-api 
  PROFILE.md ### Signal — auth model is intentional
 comment near OpenLocalStream

For #3 and #4 a suggestion is:
  Push keepalive into the dialer itself. net.Dialer has a KeepAlive          
  time.Duration field that handles this internally. Plumb a KeepAlive field
  through DialerWithResolverCache so the inner net.Dialer is constructed with   
  it:             

  stdDialer := net.Dialer{
      Resolver:  resolver.getNetResolver(),
      LocalAddr: &net.TCPAddr{IP: d.localIP},                                   
      Timeout:   d.timeout,
      KeepAlive: d.keepAlive,    // <-- new                                     
  }

**Comment by milan-zededa** [comment]
@eriknordmark Answers to those comments:
```
1. I have an integration test prepared, which covers this LPS functionality end-to-end. Therefore I do not plan to create some limited (and LPS-faking) unit tests.

2. OK, added comment

3. and 4. I added log when we fail to set keep-alive for whatever reason (should be unreachable today)

5. DNS is not persisted and the current code always sets mask; skipping this one

6. and 7. OK, added checks to prevent LPS messing with EVE

8. I didn't get this one so skipping it.
```

File: docs/CONFIG-PROPERTIES.md (hunk 1 above)
Comment (milan-zededa) [reply docs/CONFIG-PROPERTIES.md:109]: Done (and moved all these LPS interval options to a separate sub-section for better readability)

**Review by eriknordmark — APPROVED** [review]
LGTM
